Item 9093: System Integration - Purchasing Module Tickets

labkey-bpatel opened the pull request:
#### Rationale
We have completed development on the WNPRC Purchasing module, and the client has begun the testing.
These are the change requests reported by the client:
Ticket [43423](https://www.labkey.org/WNPRC/support%20tickets/issues-details.view?issueId=43423): Purchasing module - more feature requests
Ticket [43386](https://www.labkey.org/WNPRC/support%20tickets/issues-details.view?issueId=43386): Purchasing module - configure email notifications
Ticket [43385](https://www.labkey.org/WNPRC/support%20tickets/issues-details.view?issueId=43385): Purchasing module - decouple email from the purchasing requests
Ticket [43384](https://www.labkey.org/WNPRC/support%20tickets/issues-details.view?issueId=43384): Purchasing module - 'assigned to' field shows everyone
Ticket [43335](https://www.labkey.org/WNPRC/support%20tickets/issues-details.view?issueId=43335): Purchasing module - User Account Association

Spec/Testing steps [here](https://docs.google.com/document/d/1U7YjEoTSfcyjMe5TI64elol2yb3SiV2fjZJDib1uXY8/edit#).

#### Related Pull Requests
* https://github.com/LabKey/ehrModules/pull/208

#### Changes
* Filter on admins (i.e. purchase admins) minus the site admins for Assigned To dropdown on data entry form.
* Rename QC Status label from 'Order Delivered' to 'Order Received'.
* Add links for 'Request Approved' & 'Requests Ordered' on purchase admin page.
* Decouple email from the purchasing requests - log the error but let the request submission proceed.
* Add displayName, show displayName for User Accounts Associations, and data entry.
* Filter charge account on isAcceptingCharges for 'User Account Associations' and data entry page drop down.
* Make Vendor table's State and Zip columns to allow null vals for the non-US countries.

labkey-bpatel committed:
Support Ticket 43384: Purchasing module - 'assigned to' field shows everyone. Filter on admins.

labkey-bpatel committed:
Get folder admins (i.e. purchase admins) minus the site admins in the Assigned To drop down.

labkey-bpatel committed:
Ticket 43423: Purchasing module - more feature requests
- QC status drop down, Change "Order Delivered" to "Order Received"
- Add links for 'Requests Approved' & 'Requests Ordered'

labkey-bpatel committed:
Ticket 43385: Purchasing module - decouple email from the purchasing requests
- log error while sending an email but let the request submission proceed.

labkey-bpatel committed:
Add RuntimeException

labkey-bpatel committed:
Ticket 43335: Purchasing module - User Account Association
- Add displayName, show displayName for User Accounts Associations, and data entry.

labkey-bpatel committed:
Auto test updates

labkey-bpatel committed:
Auto test fix

labkey-bpatel committed:
Auto test -  change status to Order Received, change in expected email recipients

labkey-bpatel committed:
Filter charge account on isAcceptingCharges for 'User Account Associations' and data entry page drop down.

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -90,6 +94,16 @@ public WNPRC_PurchasingController()
+    {
+        //get only true folder admins, i.e. don't include site admins
+        List<User> adminUsers = SecurityManager.getUsersWithPermissions(getContainer(), Collections.singleton(AdminPermission.class));
+        List<User> siteAdminUsers = SecurityManager.getUsersWithPermissions(getContainer(), Collections.singleton(SiteAdminPermission.class));
```
This may not be quite right. It looks like it screens out all site admins, even if they're directly added as folder admins to the purchasing module. We'd want them to be included too, right?

labkey-bpatel replied:
Now changed to use role based filtering

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -183,17 +230,26 @@ public Object execute(RequestForm requestForm, BindException errors) throws Exce
+                }
             }
-            else
+            catch (RuntimeException | MessagingException | IOException | ValidationException e)
```
This seems overly broad. A failure to send emails shouldn't cause ValidationException or NPEs, should it?

labkey-bpatel replied:
Yes, however, NotificationService.get().sendMessageForRecipient throws all three exceptions

labkey-jeckels replied:
Looking at the sendMessageForRecipient, I don't really like how it's using ValidationException but catching it here seems OK. I'm more worried about the RuntimeException part of the catch - were you getting any of those thrown in your development?

labkey-jeckels replied:
I think this note got missed. I recommend removing RuntimeException from this list. And also logging the full exception instead of just the message in the _log.error() call below.

labkey-bpatel replied:
Done!

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -301,6 +357,7 @@ private void sendRequestChangeEmailNotification(String oldStatus, EmailTemplateF
                         catch (javax.mail.internet.AddressException e)
                         {
                             _log.error("Error sending line item update message to " + endUser.getEmail() , e);
+                            throw new MessagingException(e.getMessage());
```
Please pass the exception itself - chaining exceptions makes stack traces much more useful

labkey-bpatel replied:
good point, done!

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/table/WNPRC_PurchasingCustomizer.java:
```diff
@@ -33,6 +33,23 @@ else if (matches(tableInfo, "ehr_purchasing", "purchasingReceiverOverview"))
+        String name = "displayName";
+        if (aliasesTable.getColumn(name) == null)
+        {
+            SQLFragment sql = new SQLFragment("(alias ||' - '|| groupName)");
```
Will either alias or groupName ever be null?

labkey-bpatel replied:
Condition now added

labkey-bpatel committed:
Code review changes & other change requests:
- Allow state and zip to be null for vendors
- Get Folder Admin role based users instead of separately filtering out site admins.
- Handle null groupName for display

labkey-jeckels commented on WNPRC_Purchasing/resources/data/QCStatus.tsv:
```diff
@@ -3,5 +3,5 @@ Review Pending	Request is created, and its pending review	true
 Request Rejected	Request is denied or rejected	true
 Order Placed	Order is placed	true
-Order Delivered	Requested order is delivered	true
+Order Received	Requested order is delivered	true
```
Let's update the description to match.

labkey-bpatel replied:
done

labkey-jeckels commented on WNPRC_Purchasing/resources/views/purchasingAdmin.html:
```diff
@@ -4,6 +4,8 @@ <h4 style="font-family: Roboto;font-size: 14px;font-weight: bold;">Purchasing Re
 <div style="margin-left:10px;height: 25px;"><span><a class="labkey-text-link" href="<%=contextPath%><%=containerPath%>/query-executeQuery.view?schemaName=ehr_purchasing&query.queryName=myOpenRequests">My Open Requests</a></span></div>
 <div style="margin-left:10px;height: 25px;"><span><a class="labkey-text-link" href="<%=contextPath%><%=containerPath%>/query-executeQuery.view?schemaName=ehr_purchasing&query.queryName=purchasingRequestsOverviewForAdmins&query.viewName=AllOpenRequests">All Open Requests</a></span></div>
 <div style="margin-left:10px;height: 25px;"><span><a class="labkey-text-link" href="<%=contextPath%><%=containerPath%>/query-executeQuery.view?schemaName=ehr_purchasing&query.queryName=purchasingRequestsOverviewForAdmins&query.viewName=CompletedRequests">Completed Requests</a></span></div>
+<div style="margin-left:10px;height: 25px;"><span><a class="labkey-text-link" href="<%=contextPath%><%=containerPath%>/query-executeQuery.view?schemaName=ehr_purchasing&query.queryName=purchasingRequestsOverviewForAdmins&query.viewName=ApprovedRequests">Requests Approved</a></span></div>
```
Suggest making the word ordering consistent here. No strong preference on which order, but they should all be 'Requests Xed' or 'Xed Requests"

labkey-bpatel replied:
done

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -84,12 +88,36 @@
+    {
+        List<User> adminUsers = SecurityManager.getUsersWithPermissions(getContainer(), Collections.singleton(AdminPermission.class));
+        List<RoleAssignment> folderAdminGroups = getContainer().getPolicy().getAssignments().stream().filter(roleAssignment -> roleAssignment.getRole().getName().equals(FOLDER_ADMIN_ROLE)).collect(Collectors.toList());
+        int folderAdminUserId = folderAdminGroups.get(0).getUserId();
```
Won't this blow up if nobody is assigned to the folder admin role?

labkey-bpatel replied:
This code is no longer relevant. We are now handling this part the right way as per your suggestion.

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -84,12 +88,36 @@
+        List<RoleAssignment> folderAdminGroups = getContainer().getPolicy().getAssignments().stream().filter(roleAssignment -> roleAssignment.getRole().getName().equals(FOLDER_ADMIN_ROLE)).collect(Collectors.toList());
+        int folderAdminUserId = folderAdminGroups.get(0).getUserId();
+
+        List<User> folderAdmins = new ArrayList<>();
```
Instead of getting the users with permissions and reconciling against the folder admin list, I recommend iterating through the role assignments and using SecurityManager.getAllGroupMembers() to expand to a user list. Please be sure to test both direct user assignments and group assignments to the role.

labkey-bpatel replied:
Done! Thank your for suggesting this, it is much better!

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -146,6 +174,39 @@ public ModelAndView getView(Object o, BindException errors)
+                resultProperties.put("results", results);
+                return new ApiSimpleResponse(resultProperties);
+            }
+            catch (IllegalArgumentException e)
```
When does IllegalArgumentException get thrown? In general, we shouldn't be catching and handling RuntimeExceptions - we should prevent them from being thrown in the first place.

labkey-bpatel replied:
No longer, looks like it must have been from one of the earlier impl of getFolderAdmins()

labkey-bpatel replied:
Removed

labkey-bpatel committed:
Code review changes:
- Update the description for qcstate
- Make word ordering consistent for the Admin page
- Iterate through role assignments instead of reconciling against the folder admin list
- Remove <inserUrl/> from Vendors metadata
- Remove catching of IllegalArgumentException

labkey-jeckels reviewed (approved):
I think one comment on the exception handling when there's an error sending mail got missed. But no need for additional review unless you want one.

labkey-bpatel committed:
Code review change: remove RuntimeException and log the full error

labkey-bpatel committed:
Item 9093 : Allow purchase admins to add a reason for rejecting the purchasing request (#95)

* Add a field that allows purchase admins to include the reason for rejecting the request.
* Update email notification to include reason for rejection
* Update test to include reason for rejection

labkey-bpatel committed:
Merge remote-tracking branch 'origin/release21.3-SNAPSHOT' into 21.3_fb_wnprc_purchasing_issues_1

labkey-bpatel committed:
Bug fixing:
- Reordering a request should send email notification just as it is a new request.
- Reordering a request should not display 'Reason for rejection' input box.

labkey-bpatel committed:
Catch ConfigurationException as well to log the msg during request submission for when smtp server is not available.

labkey-bpatel committed:
Auto test fix
